backend/src/main/java/ai/verta/modeldb/versioning/BlobDAORdbImpl.java

```diff
@@ -412,50 +422,173 @@ boolean childContains(Set<?> list, Set<?> sublist) {
+      }
+
+      Map<String, Map.Entry<BlobExpanded, String>> locationBlobsMapCommitB =
+          getCommitBlobMapWithHash(session, internalCommitB.getRootSha(), new ArrayList<>());
+      session.getTransaction().commit();
+
+      Map<String, Map.Entry<BlobExpanded, String>> locationBlobsMapParentCommit =
+          getCommitBlobMapWithHash(session, parentCommit.getRootSha(), new ArrayList<>());
+      List<ai.verta.modeldb.versioning.BlobDiff> diffB =
+          computeDiffFromCommitMaps(locationBlobsMapCommitB, locationBlobsMapParentCommit)
+              .getDiffsList();
+      List<BlobContainer> blobContainerList =
+          getBlobContainers(diffB, locationBlobsMapCommitASimple);
+
+      final String rootSha = setBlobs(blobContainerList, new FileHasher());
+      long timeCreated = new Date().getTime();
+      List<String> parentSHAs = Arrays.asList(request.getCommitShaA(), request.getCommitShaB());
+      List<CommitEntity> parentCommits = Arrays.asList(internalCommitA, internalCommitB);
+      String mergeMessage = request.getContent().getMessage();
+      if (mergeMessage.isEmpty()) {
+        mergeMessage =
+            "Merge "
+                + request.getCommitShaA().substring(0, 7)
+                + " into "
+                + request.getCommitShaB().substring(0, 7);
+      }
+      UserInfo currentLoginUserInfo = authService.getCurrentLoginUserInfo();
+      String author = authService.getVertaIdFromUserInfo(currentLoginUserInfo);
+      final String commitSha =
+          VersioningUtils.generateCommitSHA(parentSHAs, mergeMessage, timeCreated, author, rootSha);
+
+      Commit internalCommit =
+          Commit.newBuilder()
+              .setDateCreated(timeCreated)
+              .setAuthor(author)
+              .setMessage(mergeMessage)
+              .setCommitSha(commitSha)
+              .build();
+      CommitEntity commitEntity =
+          new CommitEntity(
+              repositoryFunction.apply(session), parentCommits, internalCommit, rootSha);
+      session.saveOrUpdate(commitEntity);
+      session.getTransaction().commit();
+      return MergeRepositoryCommitsRequest.Response.newBuilder()
+          .setCommit(commitEntity.toCommitProto())
           .build();
     }
   }
 
+  private CommitEntity getCommonParent(Session session, String commitA, String commitB)
+      throws ModelDBException {
+    List<CommitEntity> parentCommitA = VersioningUtils.getParentCommits(session, commitA);
+    List<CommitEntity> parentCommitB = VersioningUtils.getParentCommits(session, commitB);
+
+    CommitEntity commonParent = null;
+    int itrA = 0;
+    int itrB = 0;
+    // TODO : this algorithm does not require all the commits to be available before starting.
+    while (itrA < parentCommitA.size() && itrB < parentCommitB.size()) {
+      CommitEntity candidateA = parentCommitA.get(itrA);
+      CommitEntity candidateB = parentCommitB.get(itrB);
```

Comment:
This (probably) either assumes we keep all commits in memory or recomputes them based on an index. Seems like a waste. We used a heap so that we can keep only the relevant ones there.

Reply:
assumes we keep all `parent` commits in memory.
It does not recompute them.
Agreed that it is not optimized but the heap requirement seemed unnecessary.